File: core/infra/platform/windows.go
// Package platform provides OS-specific abstractions for ELMOS operations.
// This file implements the Windows platform by delegating all operations to
// a WSL2 (Windows Subsystem for Linux) environment via wsl.exe.
//
// Requirements:
//   - WSL2 installed with a Debian/Ubuntu distribution
//   - wsl.exe available on the system PATH
package platform

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/NguyenTrongPhuc552003/elmos/core/infra/executor"
)

// windowsPlatform implements Platform for Windows hosts via WSL2 delegation.
type windowsPlatform struct {
	exec     executor.Executor
	diskImg  *windowsDiskImage
	packages *windowsPackages
	paths    *windowsPaths
}

func newWindowsPlatform(exec executor.Executor) *windowsPlatform {
	if exec == nil {
		exec = executor.NewShellExecutor()
	}
	return &windowsPlatform{
		exec:     exec,
		diskImg:  &windowsDiskImage{exec: exec},
		packages: &windowsPackages{exec: exec},
		paths:    &windowsPaths{},
	}
}

func (w *windowsPlatform) Name() string                { return "windows-wsl2" }
func (w *windowsPlatform) DiskImage() DiskImageManager { return w.diskImg }
func (w *windowsPlatform) Packages() PackageManager    { return w.packages }
func (w *windowsPlatform) Paths() PathProvider         { return w.paths }

func (w *windowsPlatform) SetExecutor(exec executor.Executor) {
	w.exec = exec
	w.diskImg.exec = exec
	w.packages.exec = exec
}

// --- DiskImageManager (Windows: delegates to WSL2) ---

type windowsDiskImage struct{ exec executor.Executor }

// Create creates a sparse image inside the WSL2 filesystem using fallocate.
func (d *windowsDiskImage) Create(ctx context.Context, imagePath string, sizeGB int) error {
	wslPath := windowsToWSLPath(imagePath)
	sizeBytes := fmt.Sprintf("%d", int64(sizeGB)*1024*1024*1024)
	return d.exec.Run(ctx, "wsl.exe", "--", "fallocate", "-l", sizeBytes, wslPath)
}

// Mount mounts the image inside WSL2 at /mnt/elmos/<name>.
func (d *windowsDiskImage) Mount(ctx context.Context, imagePath string) (string, error) {
	wslPath := windowsToWSLPath(imagePath)
	name := strings.TrimSuffix(filepath.Base(imagePath), filepath.Ext(imagePath))
	mountPoint := fmt.Sprintf("/mnt/elmos/%s", name)

	// Ensure mount directory exists
	if err := d.exec.Run(ctx, "wsl.exe", "--", "mkdir", "-p", mountPoint); err != nil {
		return "", fmt.Errorf("wsl mkdir: %w", err)
	}

	// Allocate loop device
	loopDevice, err := d.exec.Output(ctx, "wsl.exe", "--", "losetup", "--find", "--show", wslPath)
	if err != nil {
		return "", fmt.Errorf("wsl losetup: %w", err)
	}
	loopDev := strings.TrimSpace(string(loopDevice))

	// Mount the loop device
	if err := d.exec.Run(ctx, "wsl.exe", "--", "mount", loopDev, mountPoint); err != nil {
		// Cleanup: detach loop device on mount failure
		_ = d.exec.Run(ctx, "wsl.exe", "--", "losetup", "-d", loopDev)
		return "", fmt.Errorf("wsl mount: %w", err)
	}

	return mountPoint, nil
}

// Unmount unmounts inside WSL2.
func (d *windowsDiskImage) Unmount(ctx context.Context, mountPoint string) error {
	// Unmount first
	if err := d.exec.Run(ctx, "wsl.exe", "--", "umount", mountPoint); err != nil {
		return fmt.Errorf("wsl umount: %w", err)
	}

	// Find and detach loop device
	loopInfo, err := d.exec.Output(ctx, "wsl.exe", "--", "losetup", "-j", mountPoint)
	if err == nil && len(loopInfo) > 0 {
		// Extract loop device name (format: "/dev/loop0: ...")
		parts := strings.SplitN(string(loopInfo), ":", 2)
		if len(parts) > 0 {
			loopDev := strings.TrimSpace(parts[0])
			_ = d.exec.Run(ctx, "wsl.exe", "--", "losetup", "-d", loopDev)
		}
	}

	return nil
}

// IsMounted checks /proc/mounts inside WSL2.
func (d *windowsDiskImage) IsMounted(ctx context.Context, name string) (bool, string, error) {
	target := fmt.Sprintf("/mnt/elmos/%s", name)

	// Use grep to check /proc/mounts directly
	err := d.exec.Run(ctx, "wsl.exe", "--", "grep", "-q", " "+target+" ", "/proc/mounts")
	if err != nil {
		// grep returns non-zero exit code if not found
		return false, "", nil
	}

	return true, target, nil
}

// --- PackageManager (Windows: apt-get inside WSL2) ---

type windowsPackages struct{ exec executor.Executor }

func (p *windowsPackages) IsInstalled(pkg string) bool {
	out, err := p.exec.Output(context.Background(), "wsl.exe", "--", "dpkg", "-s", pkg)
	if err != nil {
		return false
	}
	return strings.Contains(string(out), "Status: install ok installed")
}

func (p *windowsPackages) Install(ctx context.Context, pkg string) error {
	return p.exec.Run(ctx, "wsl.exe", "--", "apt-get", "install", "-y", pkg)
}

func (p *windowsPackages) ListInstalled() ([]string, error) {
	out, err := p.exec.Output(context.Background(), "wsl.exe", "--", "dpkg", "--get-selections")
	if err != nil {
		return nil, fmt.Errorf("wsl dpkg --get-selections: %w", err)
	}
	var result []string
	for _, line := range strings.Split(string(out), "\n") {
		if line != "" {
			result = append(result, strings.Fields(line)[0])
		}
	}
	return result, nil
}

func (p *windowsPackages) GetBinPath(pkg string) string {
	out, err := p.exec.Output(context.Background(), "wsl.exe", "--", "which", pkg)
	if err != nil || len(out) == 0 {
		return ""
	}
	return filepath.Dir(strings.TrimSpace(string(out)))
}

func (p *windowsPackages) GetLibPath(_ string) string     { return "/usr/lib" }
func (p *windowsPackages) GetIncludePath(_ string) string { return "/usr/include" }

// --- PathProvider (Windows) ---

type windowsPaths struct{}

// WorkspaceRoot returns the UNC path to the workspace directory in WSL2.
// This allows Windows tools to access the workspace if needed.
func (p *windowsPaths) WorkspaceRoot(name string) string {
	return fmt.Sprintf("\\\\wsl$\\Ubuntu\\mnt\\elmos\\%s", name)
}

func (p *windowsPaths) CacheDir() string {
	homeDir, err := os.UserHomeDir()
	if err != nil {
		return filepath.Join("~", ".elmos")
	}
	return filepath.Join(homeDir, ".elmos")
}

func (p *windowsPaths) ToolchainDir() string {
	homeDir, err := os.UserHomeDir()
	if err != nil {
		return filepath.Join("~", "x-tools")
	}
	return filepath.Join(homeDir, "x-tools")
}

// windowsToWSLPath converts a Windows-style path to a WSL2 /mnt/... path.
// e.g. C:\Users\foo\elmos.img → /mnt/c/Users/foo/elmos.img
func windowsToWSLPath(winPath string) string {
	if len(winPath) >= 2 && winPath[1] == ':' {
		drive := strings.ToLower(string(winPath[0]))
		rest := strings.ReplaceAll(winPath[2:], "\\", "/")
		return fmt.Sprintf("/mnt/%s%s", drive, rest)
	}
	return strings.ReplaceAll(winPath, "\\", "/")
}

File: src/infra/platform/windows.cpp
// ============================================================================
// platform/windows.cpp — Windows (WSL2) platform implementation
// ============================================================================

#ifdef ELMOS_PLATFORM_WINDOWS

#include "windows.hpp"

#include <algorithm>
#include <cstdlib>
#include <filesystem>
#include <sstream>

namespace elmos::infra::platform {

auto windows_to_wsl_path(const std::string& win_path) -> std::string {
    if (win_path.size() >= 2 && win_path[1] == ':') {
        char drive = static_cast<char>(std::tolower(static_cast<unsigned char>(win_path[0])));
        auto rest = win_path.substr(2);
        std::replace(rest.begin(), rest.end(), '\\', '/');
        return "/mnt/" + std::string(1, drive) + rest;
    }
    auto result = win_path;
    std::replace(result.begin(), result.end(), '\\', '/');
    return result;
}

// --- WindowsPlatform ---

WindowsPlatform::WindowsPlatform(executor::Executor* exec) : exec_(exec) {
    disk_img_ = std::make_unique<WindowsDiskImage>(exec_);
    packages_ = std::make_unique<WindowsPackages>(exec_);
}

void WindowsPlatform::set_executor(executor::Executor* exec) {
    exec_ = exec;
    disk_img_->set_executor(exec);
    packages_->set_executor(exec);
}

// --- WindowsDiskImage ---

auto WindowsDiskImage::create(std::stop_token token, const std::string& image_path, int size_gb)
    -> VoidResult {
    auto wsl_path = windows_to_wsl_path(image_path);
    auto size_bytes = std::to_string(static_cast<int64_t>(size_gb) * 1024 * 1024 * 1024);
    return exec_->run(token, "wsl.exe", {"--", "fallocate", "-l", size_bytes, wsl_path});
}

auto WindowsDiskImage::mount(std::stop_token token, const std::string& image_path)
    -> Result<std::string> {
    auto wsl_path = windows_to_wsl_path(image_path);
    auto name = std::filesystem::path(image_path).stem().string();
    auto mount_point = "/mnt/elmos/" + name;

    auto r1 = exec_->run(token, "wsl.exe", {"--", "mkdir", "-p", mount_point});
    if (!r1)
        return make_error(Error::wrap("wsl mkdir", r1.error()));

    auto out = exec_->output(token, "wsl.exe", {"--", "losetup", "--find", "--show", wsl_path});
    if (!out)
        return make_error(Error::wrap("wsl losetup", out.error()));

    auto loop_dev = *out;
    while (!loop_dev.empty() && (loop_dev.back() == '\n' || loop_dev.back() == ' ')) {
        loop_dev.pop_back();
    }

    auto r2 = exec_->run(token, "wsl.exe", {"--", "mount", loop_dev, mount_point});
    if (!r2) {
        exec_->run(token, "wsl.exe", {"--", "losetup", "-d", loop_dev});
        return make_error(Error::wrap("wsl mount", r2.error()));
    }

    return mount_point;
}

auto WindowsDiskImage::unmount(std::stop_token token, const std::string& mount_point)
    -> VoidResult {
    auto r = exec_->run(token, "wsl.exe", {"--", "umount", mount_point});
    if (!r)
        return make_error(Error::wrap("wsl umount", r.error()));

    auto out = exec_->output(token, "wsl.exe", {"--", "losetup", "-j", mount_point});
    if (out && !out->empty()) {
        auto colon = out->find(':');
        if (colon != std::string::npos) {
            auto loop_dev = out->substr(0, colon);
            while (!loop_dev.empty() && loop_dev.front() == ' ')
                loop_dev.erase(loop_dev.begin());
            exec_->run(token, "wsl.exe", {"--", "losetup", "-d", loop_dev});
        }
    }
    return {};
}

auto WindowsDiskImage::is_mounted(std::stop_token token, const std::string& name)
    -> Result<MountStatus> {
    auto target = "/mnt/elmos/" + name;
    auto r = exec_->run(token, "wsl.exe", {"--", "grep", "-q", " " + target + " ", "/proc/mounts"});
    if (!r)
        return MountStatus{false, ""};
    return MountStatus{true, target};
}

// --- WindowsPackages ---

auto WindowsPackages::is_installed(const std::string& pkg) -> bool {
    std::stop_source ss;
    auto out = exec_->output(ss.get_token(), "wsl.exe", {"--", "dpkg", "-s", pkg});
    return out.has_value() && out->find("Status: install ok installed") != std::string::npos;
}

auto WindowsPackages::install(std::stop_token token, const std::string& pkg) -> VoidResult {
    return exec_->run(token, "wsl.exe", {"--", "apt-get", "install", "-y", pkg});
}

auto WindowsPackages::list_installed() -> Result<std::vector<std::string>> {
    std::stop_source ss;
    auto out = exec_->output(ss.get_token(), "wsl.exe", {"--", "dpkg", "--get-selections"});
    if (!out)
        return make_error(Error::wrap("wsl dpkg", out.error()));

    std::vector<std::string> result;
    std::istringstream stream(*out);
    std::string line;
    while (std::getline(stream, line)) {
        if (!line.empty()) {
            auto space = line.find_first_of(" \t");
            result.push_back(space != std::string::npos ? line.substr(0, space) : line);
        }
    }
    return result;
}

auto WindowsPackages::get_bin_path(const std::string& pkg) -> std::string {
    std::stop_source ss;
    auto out = exec_->output(ss.get_token(), "wsl.exe", {"--", "which", pkg});
    if (!out || out->empty())
        return "";
    auto path = *out;
    while (!path.empty() && (path.back() == '\n' || path.back() == ' '))
        path.pop_back();
    return std::filesystem::path(path).parent_path().string();
}

auto WindowsPackages::get_lib_path(const std::string& /*pkg*/) -> std::string {
    return "/usr/lib";
}

auto WindowsPackages::get_include_path(const std::string& /*pkg*/) -> std::string {
    return "/usr/include";
}

// --- WindowsPaths ---

auto WindowsPaths::workspace_root(const std::string& name) -> std::string {
    return "\\\\wsl$\\Ubuntu\\mnt\\elmos\\" + name;
}

auto WindowsPaths::cache_dir() -> std::string {
    if (const char* home = std::getenv("USERPROFILE")) {
        return std::string(home) + "\\.elmos";
    }
    return "~\\.elmos";
}

auto WindowsPaths::toolchain_dir() -> std::string {
    if (const char* home = std::getenv("USERPROFILE")) {
        return std::string(home) + "\\x-tools";
    }
    return "~\\x-tools";
}

}  // namespace elmos::infra::platform

#endif  // ELMOS_PLATFORM_WINDOWS
